dbms/src/Interpreters/sortBlock.cpp

```diff
@@ -71,48 +299,101 @@ struct PartialSortingLess
         : columns(columns_)
     {}
 
-    bool operator()(size_t a, size_t b) const
+    ALWAYS_INLINE inline bool operator()(size_t a, size_t b) const
     {
+        int res = 0;
         for (const auto & column : columns)
         {
-            int res = column.second.direction * column.first->compareAt(a, b, *column.first, column.second.nulls_direction);
-            if (res < 0)
-                return true;
-            else if (res > 0)
-                return false;
+            res = column.second.direction * column.first->compareAt(a, b, *column.first, column.second.nulls_direction);
+            if (res)
+                break;
         }
-        return false;
+        return res < 0;
     }
 };
 
 struct PartialSortingLessWithCollation
 {
-    const ColumnsWithSortDescriptions & columns;
+    const CollatorDesc & collator_desc;
```

Comment:
Using `CollatorDesc` instead of `ColumnsWithSortDescriptions` seems does not make too much difference?

Reply:
`PartialSortingLessWithCollation` uses member variables `has_collation` and `need_collations` of `CollatorDesc `